example/bwmf/bwmf_task.go

```diff
@@ -168,19 +288,68 @@ func (t *bwmfTask) doDataReady(ctx context.Context, fromID uint64, method string
 
 // These two function carry out actual optimization.
 func (t *bwmfTask) updateDShard(ctx context.Context) {
-	select {
-	case <-time.After(time.Duration(rand.Intn(3)) * time.Second):
-	case <-ctx.Done():
-		return
+
+	if t.dims.N == -1 {
+		t.dims.N = 0
+		for _, m := range t.peerShards {
+			t.dims.N += len(m.Row)
+		}
+		wh := make([][]float32, t.dims.N)
+		for i, _ := range wh {
+			wh[i] = make([]float32, t.dims.m)
+		}
+		t.dLoss = &KLDivLoss{
+			V:  t.rowShard,
+			WH: wh,
+			m:  t.dims.N,
+			n:  t.dims.m,
+			k:  t.dims.k,
+		}
+	}
+
+	t.dLoss.W = NewBlocksParameter(&t.peerShards)
+	param := NewSingleBlockParameter(t.dShard)
+
+	ok := t.optimizer.Minimize(t.dLoss, t.stopCriteria, param)
+	if !ok {
+		t.logger.Panicf("Failed minimizing over dShard.")
+		// handle re-run
+	}
+
+	t.curIter += 1
+	if t.curIter >= t.numIters {
+		t.exitChan <- &event{}
```

Comment:
One fallacy here is that when we send an event to exitChan, we should not pass any event to updateChan. I'll fix this later.